worker/task_manager.go

```diff
@@ -0,0 +1,263 @@
+		return nil, err
+	}
+
+	m.pluginManager = pm
+	return m, nil
+}
+
+// Start will initiliaze a polling cycle for tasks and spawn goroutines to
+// execute units of work that has been claimed.
+func (m *Manager) Start(stop <-chan struct{}, done chan struct{}) {
+	m.log.Info("Polling for tasks every %d seconds\n", m.interval)
+	doWork := time.NewTicker(time.Duration(m.interval) * time.Second)
+	for {
+		select {
+		case <-stop:
+			doWork.Stop()
+			go m.Stop()
+		case <-doWork.C:
+			n := math.Max(float64(m.maxCapacity-len(m.tasks)), 0)
+			m.claimWork(int(n))
+		case <-m.done:
+			close(done)
+			return
+		}
+	}
+}
+
+func (m *Manager) Stop() {
+	defer close(m.done)
+	// Do interesting things
+	// Loop over all running tasks, call cancel
+	return
+}
+
+func (m *Manager) claimWork(ntasks int) {
+	if ntasks == 0 {
+		return
+	}
+
+	claims := m.queue.ClaimWork(ntasks)
+	for _, c := range claims {
+		go m.runTask(c)
+	}
+}
+
+func (m *Manager) runTask(task *TaskRun) error {
+	log := m.log.WithFields(logrus.Fields{
+		"taskId": task.TaskId,
+		"runId":  task.RunId,
+	})
+	task.log = log
+	log.Info("Running Task")
+
+	err := m.registerTask(task)
+	done := make(chan struct{})
+	defer func() {
+		m.deregisterTask(task)
+	}()
+
+	if err != nil {
+		log.WithField("error", err.Error()).Warn("Could not register task")
+		return err
+	}
+
+	tp := m.environment.TemporaryStorage.NewFilePath()
+	task.context, task.controller, err = runtime.NewTaskContext(tp)
+	if err != nil {
+		log.WithField("error", err.Error()).Warn("Could not create task context")
+		return err
+	}
+
+	jsonPayload := map[string]json.RawMessage{}
+	if err := json.Unmarshal(task.Definition.Payload, &jsonPayload); err != nil {
+		log.WithField("error", err.Error()).Warn("Could not parse task payload")
+		return err
+	}
+
+	p, err := m.engine.PayloadSchema().Parse(jsonPayload)
+	if err != nil {
+		log.WithField("error", err.Error()).Warn("Payload validation failed: %s", task.Definition.Payload)
+		return err
+	}
+	task.payload = p
+
+	ps, err := m.pluginManager.PayloadSchema()
+	if err != nil {
+		log.WithField("error", err.Error()).Warn("Could not retrieve plugin payload schemas")
+		return err
+	}
+
+	pluginPayload, err := ps.Parse(jsonPayload)
+	if err != nil {
+		log.WithField("error", err.Error()).Warn("Plugin payload validation failed: %s", task.Definition.Payload)
+		return err
+	}
+
+	popts := plugins.TaskPluginOptions{TaskInfo: &runtime.TaskInfo{}, Payload: pluginPayload}
+	taskPlugins, err := m.pluginManager.NewTaskPlugin(popts)
+	if err != nil {
+		log.WithField("error", err.Error()).Warn("Could not create task plugins")
+		return err
+	}
+
+	task.plugin = taskPlugins
+
+	// each method can return error channel, and the next stage can consume it
+	prepare, errc := task.Prepare(m.engine)
+	build, errc := task.Build(prepare, done, errc)
+	run, errc := task.Run(build, done, errc)
+	stop, errc := task.Stop(run, done, errc)
+	finish, errc := task.Finish(stop, done, errc)
```

Comment:
Yea so this attempt was to see what a flow of stages looked like, where a task will always end up in the finished stages regardless of what happened in any of the intermediary stages.  It's probably completely crazy, just something I was experimenting with